File: frontend/src/lib/queries/churn.ts
'use client'

import { useMutation } from '@tanstack/react-query'
import { useApi } from '@/lib/api/use-api'

export type TopDriver = {
  feature: string
  shap_value: number
  direction: 'increases_risk' | 'decreases_risk'
}

export type AtRiskCustomer = {
  customer: string
  name: string
  churn_probability: number
  risk_tier: string
  recency_days: number
  frequency: number
  monetary: number | null
  expected_loss: number | null
  top_drivers: TopDriver[]
}

export type FeatureImportance = {
  feature: string
  importance: number
}

export type ChurnRunResult = {
  available: boolean
  reason: string | null
  horizon_days: number | null
  requested_horizon_days: number | null
  snapshot_date: string | null
  cutoff_date: string | null
  training_cutoffs: string[]
  model: Record<string, any>
  feature_importance: FeatureImportance[]
  shap_global_importance: FeatureImportance[]
  risk_distribution: Record<string, number>
  customers_scored: number
  revenue_at_risk: number | null
  expected_revenue_at_risk: number | null
  at_risk_customers: AtRiskCustomer[]
  at_risk_ranked_by: string
  metadata: Record<string, any>
  customer_scores: Record<string, number> | null
}

export function useRunChurn(projectId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (input: { horizon_days: number; top_n: number }) =>
      call<ChurnRunResult>(`/api/v1/projects/${projectId}/churn/run`, { method: 'POST', json: input }),
  })
}

export function useRetentionPlan(projectId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (input: { churn_result: ChurnRunResult; business_context?: string }) =>
      call<{ report_md: string }>(`/api/v1/projects/${projectId}/churn/retention-plan`, {
        method: 'POST',
        json: input,
      }),
  })
}

File: backend/app/schemas/churn.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ChurnRunRequest(BaseModel):
    horizon_days: int = 90
    top_n: int = 200


class TopDriver(BaseModel):
    feature: str
    shap_value: float
    direction: str


class AtRiskCustomer(BaseModel):
    customer: str
    name: str = ""
    churn_probability: float
    risk_tier: str
    recency_days: int
    frequency: int
    monetary: float | None = None
    expected_loss: float | None = None
    top_drivers: list[TopDriver] = Field(default_factory=list)


class FeatureImportance(BaseModel):
    feature: str
    importance: float


class ChurnRunResponse(BaseModel):
    available: bool
    reason: str | None = None
    horizon_days: int | None = None
    requested_horizon_days: int | None = None
    snapshot_date: str | None = None
    cutoff_date: str | None = None
    training_cutoffs: list[str] = Field(default_factory=list)
    model: dict[str, Any] = Field(default_factory=dict)
    feature_importance: list[FeatureImportance] = Field(default_factory=list)
    shap_global_importance: list[FeatureImportance] = Field(default_factory=list)
    risk_distribution: dict[str, int] = Field(default_factory=dict)
    customers_scored: int = 0
    revenue_at_risk: float | None = None
    expected_revenue_at_risk: float | None = None
    at_risk_customers: list[AtRiskCustomer] = Field(default_factory=list)
    at_risk_ranked_by: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)
    # Full per-customer churn-probability map, used only to cross-link into
    # the Marketing agent's RFM-segment churn breakdown — not for display.
    customer_scores: dict[str, float] | None = None


class RetentionPlanRequest(BaseModel):
    churn_result: dict[str, Any]
    business_context: str = ""
    model: str | None = None


class RetentionPlanResponse(BaseModel):
    report_md: str
